Code/Engine/InputManager.h

```diff
@@ -5,100 +5,140 @@
+         * @param key scancode
+         * @return 
+        */
+        bool GetKey(SDL_Scancode key);
 
         /**
-         * @brief Saves the down state of a key.
-         * @param key
-         * @param pressed
+         * @brief Returns true if a keyboard key is pressed during the current frame
+         * @param key scancode
+         * @return
         */
-        void UpdateKeyboardState(SDL_Scancode key, bool pressed)
-        {
-            std::map<SDL_Scancode, bool>::iterator iter = KeyboardStates.find(key);
-            if (iter != KeyboardStates.end())
-            {
-                iter->second = pressed;
-                return;
-            }
-            KeyboardStates.insert({ key, pressed });
-        }
+        bool GetKeyPressed(SDL_Scancode key);
 
         /**
-         * @brief Returns true if a key is down in the previous frame.
-         * @param key
+         * @brief Returns true if a keyboard key is released during the current frame
+         * @param key scancode
          * @return
         */
-        bool GetKeyboardState(SDL_Scancode key)
-        {
-            std::map<SDL_Scancode, bool>::iterator iter = KeyboardStates.find(key);
-            if (iter != KeyboardStates.end())
-            {
-                return iter->second;
-            }
-            return false;
+        bool GetKeyReleased(SDL_Scancode key);
+
+
+        // Mouse Events
+        SDL_Point MousePosition;
+        bool mouseState[6];
+        int  mouseUpdateFrames[6];
+
+        /**
+         * @brief Catches mouse button events and updates their states
+         * @param event
+        */
+        void CatchMouseEvents(SDL_Event& event);
+        
+        /**
+         * @brief Returns true if a mouse button is held.
+         * @param button mouse button id
+         * @return
+        */
+        bool GetMouseButton(int button);
+
+        /**
+         * @brief Returns true if a mouse button is pressed during the current frame.
+         * @param button mouse button id
+         * @return
+        */
+        bool GetMouseButtonPressed(int button);
+
+        /**
+         * @brief Returns true if a mouse button is released during the current frame.
+         * @param button mouse button id
+         * @return
+        */
+        bool GetMouseButtonReleased(int button); 
+
+        void SetMousePosition(int x, int y) {
+            MousePosition.x = x;
+            MousePosition.y = y;
         }
+        SDL_Point GetMousePosition(){ return MousePosition; };
+
 
         // Controller Events
-        bool ControllerIsActive = false;
-        void OnControllerConnected(SDL_ControllerDeviceEvent& e);
-        void OnControllerDisconnected(SDL_ControllerDeviceEvent& e);
+        SDL_GameController* controller = nullptr;
+        std::unordered_map<int, bool> buttonState;
+        std::unordered_map<int, int>  buttonUpdateFrames;
+        float ControllerDeadZone;
+        bool  ControllerIsActive;
```

Comment:
Last one 😅 should this be public, do we want another class to be able to disable a controller other than the `InputManager`?